src/driver/amdxdna/ve2_hwctx.c

```diff
@@ -678,33 +825,57 @@ static int ve2_update_handshake_pkt(struct amdxdna_ctx *hwctx, u64 paddr, u8 buf
 	return ret;
 }
 
-int ve2_hwctx_config(struct amdxdna_ctx *hwctx, u32 type, u64 mdata_hdl, void *buf, u32 size)
+static int ve2_hwctx_config_op_timeout(struct amdxdna_ctx *hwctx, u32 op_timeout)
+{
+	struct device *aie_dev = hwctx->priv->aie_part;
+	struct handshake hs = { 0 };
+	int ret;
+
+	hs.opcode_timeout_config = op_timeout;
+	ret = aie_partition_write_privileged_mem(aie_dev, CERT_HANDSHAKE_OFF(hwctx->start_col) +
+						 offsetof(struct handshake, opcode_timeout_config),
+						 sizeof(u32), (void *)&hs.opcode_timeout_config);
+	return ret;
+}
+
+static struct fw_buffer_metadata *get_fwbuf_metadata_hdl(struct amdxdna_client *client,
+							 u64 mdata_hdl)
+{
+	struct amdxdna_gem_obj *mdata_abo;
+	struct fw_buffer_metadata *mdata;
+
+	mdata_abo = amdxdna_gem_get_obj(client, mdata_hdl, AMDXDNA_BO_DEV);
+	if (!mdata_abo || !mdata_abo->mem.kva)
+		return NULL;
+
+	mdata = (struct fw_buffer_metadata *)(mdata_abo->mem.kva);
+	if (!mdata)
+		amdxdna_gem_put_obj(mdata_abo);
+
+	return mdata;
```

Comment:
Where to put mdata_abo?

Reply:
amdxdna_gem_put_obj(mdata_abo) will be called in caller function at line 735

Reply:
This looks incorrect.  It should be  "*mdata_abo = abo" before function return success. And in general, please avoid allocate inside a function and release outside if possible.